### PR TITLE
fix(server, vue-app): address encoding issues with query params

### LINKED ISSUE #7600: % in URL param results in incorrect behaviour
### Version

[v2.12.2](https://github.com/nuxt.js/releases/tag/v2.12.2)

### Reproduction link

[https://32kd4.sse.codesandbox.io/](https://32kd4.sse.codesandbox.io/)

### Steps to reproduce

* Go to https://32kd4.sse.codesandbox.io/
* Click on link to "A working page" - see that the page displays correctly
* Click on link to "An invalid page" - see a "This page could not be found"
* Go back and click on link to "A broken page" - see an error page with the message "URI malformed"
* Go back and click on link to "A differently broken page" - see an ngnix error page 

### What is expected ?

The three non-working links should all display the "This page could not be found" page

### What is actually happening?

* A URL param which cannot be decoded throws a URI malformed error
* A URL param which is a single % causes a more severe error

### Additional comments?

See also https://github.com/nuxt/nuxt.js/issues/4977

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c10841">#c10841</a>)</em></sub></div>

### LINKED ISSUE #8680: URL params format error
### Versions

- nuxt@2.14.12

### Steps to reproduce

Open url like [/search/%2F](/search/%2F)

### What is Expected?

Throw error "TypeError: Cannot read property '_isDestroyed' of undefined"

### Temporary processing

Return to the version nuxt@2.14.3 , and there is no such problem.

By the way, the URL with "%25" will throw error "URIError: URI malformed", 
and I have to using a middleware translate to “%2525”.

### LINKED ISSUE #8690: Query parameter decoding error in server rendering
### Versions

- nuxt: 2.14.12
- node: 14.15.0

### Reproduction
Query parameter decoding error in SSR created hook and asyncData hook

### Steps to reproduce
https://vvw9q.sse.codesandbox.io/?str=%26&str2=%2526

### What is Expected?
Output:
```
In the asyncData hook on the server
str: &
str2: %26
In the mounted hook of the client
str: &
str2: %26
```

### What is actually happening?
Output:
```
In the asyncData hook on the server
str: &
str2: &
In the mounted hook of the client
str: &
str2: %26
```

### LINKED ISSUE #8692: '%20' caractere turning into '%2b'
### Versions

- nuxt: v2.14.12
- node: v15.2.0

### Reproduction

https://7yf7h.sse.codesandbox.io/?email=some%20aemail.com

### What is Expected?

output: some aemail.com

### What is actually happening?

output: some+aemail.com

### PR DESCRIPTION
**Pull request by danielroe**
## Types of changes
- [x] Bug fix (a non-breaking change which fixes an issue)

## Description
* needs to be matched with https://github.com/nuxt-contrib/ufo/pull/11
* note the change in behaviour with the encoding test
* closes #8692, closes #8690, closes #8680 and closes #7600.

## Checklist:
- [x] I have added tests to cover my changes
- [x] All new and existing tests are passing

### CODE REVIEW

File: test/dev/encoding.test.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ describe('encoding', () => {
   })
 
   test('/ö/dynamic?q=food,coffee (encodeURIComponent)', async () => {
-    const { body: response } = await rp(url('/ö/dynamic?q=food%252Ccoffee'))
+    const { body: response } = await rp(url('/ö/dynamic?q=food,coffee'))
```
Comment (danielroe) [inline test/dev/encoding.test.js:25]: `url` encodes this already - and I think we shouldn't be testing for double-decoding

Comment (pi0) [reply test/dev/encoding.test.js:25]: Context: #8429 / https://github.com/nuxt/nuxt.js/pull/8430

File: packages/server/src/middleware/nuxt.js [hunk 2]
```diff
@@ -9,7 +10,7 @@ export default ({ options, nuxt, renderRoute, resources }) => async function nux
   const context = getContext(req, res)
 
   try {
-    const url = decodeURI(req.url)
+    const url = normalizeURL(req.url)
```
Comment (pi0) [inline packages/server/src/middleware/nuxt.js:13]: normalizeURL _encodes_

File: test/dev/encoding.test.js (hunk 1 above)
Comment (danielroe) [reply test/dev/encoding.test.js:25]: Added tests to cover the issues reported in associated issues.

**Comment by pi0** [comment]
@danielroe Seems `normalizeURL` is already fixing encoding issues for cases in #7600 (`/%c` ~> `/%25c` and `'/%25'`) which fixes issue when using direct+invalid link. Is there any other condition not handled?

**Comment by danielroe** [comment]
@pi0 Internally I did fix that case but it involved changes across more of the code base so opted not to submit in this PR - the devMiddleware in particular would need to be changed.

**Review by pi0 — APPROVED** [review]

**Comment by leochen-g** [comment]
nuxt version is ` 2.15.8`

It seems also have this problem. You can try this dynamic params `司机指南-优步司机、滴滴司机端注册�%B`.  ` URI malformed` will also appear.

**Pull request merged**